nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,344 @@
+    throw nav2_core::ControllerException("Unable to lock node!");
+  }
+
+  costmap_ros_ = costmap_ros;
+  tf_buffer_ = tf;
+  plugin_name_ = name;
+  logger_ = node->get_logger();
+
+  // Handles storage and dynamic configuration of parameters.
+  // Returns pointer to data current param settings.
+  param_handler_ = std::make_unique<ParameterHandler>(
+    node, plugin_name_, logger_,
+    costmap_ros_->getCostmap()->getSizeInMetersX());
+  params_ = param_handler_->getParams();
+
+  // Handles global path transformations
+  path_handler_ = std::make_unique<PathHandler>(
+    tf2::durationFromSec(params_->transform_tolerance), tf_buffer_, costmap_ros_);
+
+  // Handles the control law to generate the velocity commands
+  control_law_ = std::make_unique<SmoothControlLaw>(
+    params_->k_phi, params_->k_delta, params_->beta, params_->lambda, params_->slowdown_radius,
+    params_->v_linear_min, params_->v_linear_max, params_->v_angular_max);
+
+  // Initialize footprint collision checker
+  collision_checker_ = std::make_unique<nav2_costmap_2d::
+      FootprintCollisionChecker<nav2_costmap_2d::Costmap2D *>>(costmap_ros_->getCostmap());
+
+  // Publishers
+  transformed_plan_pub_ = node->create_publisher<nav_msgs::msg::Path>("transformed_global_plan", 1);
+  local_plan_pub_ = node->create_publisher<nav_msgs::msg::Path>("local_plan", 1);
+  motion_target_pub_ = node->create_publisher<geometry_msgs::msg::PointStamped>("motion_target", 1);
+  slowdown_pub_ = node->create_publisher<visualization_msgs::msg::Marker>("slowdown", 1);
+
+  RCLCPP_INFO(logger_, "Configured Graceful Motion Controller: %s", plugin_name_.c_str());
+}
+
+void GracefulMotionController::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_.reset();
+  local_plan_pub_.reset();
+  motion_target_pub_.reset();
+  slowdown_pub_.reset();
+}
+
+void GracefulMotionController::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_->on_activate();
+  local_plan_pub_->on_activate();
+  motion_target_pub_->on_activate();
+  slowdown_pub_->on_activate();
+}
+
+void GracefulMotionController::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating controller: %s of type graceful_motion_controller::GracefulMotionController",
+    plugin_name_.c_str());
+  transformed_plan_pub_->on_deactivate();
+  local_plan_pub_->on_deactivate();
+  motion_target_pub_->on_deactivate();
+  slowdown_pub_->on_deactivate();
+}
+
+geometry_msgs::msg::TwistStamped GracefulMotionController::computeVelocityCommands(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const geometry_msgs::msg::Twist & /*velocity*/,
+  nav2_core::GoalChecker * goal_checker)
+{
+  std::lock_guard<std::mutex> param_lock(param_handler_->getMutex());
```

Comment:
Ah I see, so you lock the param mutex here. Ok. I think that makes sense, but how does the change of the K values and other values of the control law instance get updated?

Reply:
The control law parameters only can be changed when you read the parameters at the start because they are passed to the control_law class when the node is configured. Lines 50-52 of nav2_graceful_motion_controller.cpp

I can add those values to the OnSetParametersCallback and then set the control law with the new values.

Reply:
That would be great!